File: CT-Viewer-Native/CoreNative/CubeManager.cpp
#include "stdafx.h"
#include "CubeManager.h"
#include <cstring>

CoreNative::CubeManager::CubeManager()
{
	reader = new CoreNative::CTXReader();
	cubes = new std::vector<CoreNative::ImageCube*>;
};

CoreNative::CubeManager::~CubeManager()
{
	delete reader;
	if(!cubes->empty())
	{
		for(unsigned int i=0; i<cubes->size(); i++)
		{
			CoreNative::ImageCube* tempCube = cubes->back();
			cubes->pop_back();
			delete tempCube;
		}
	}
	delete cubes;
};

CoreNative::ImageCube* CoreNative::CubeManager::GetImageCube(const char* imageID)
{
	if(!cubes->empty())
	{
		for(unsigned int i=0; i<cubes->size(); i++)
		{
			CoreNative::ImageCube* tempCube = cubes->at(i);
			if( strncmp(imageID,tempCube->GetImageID(), strlen(imageID)) == 0 )
			{
				return tempCube;
			}
		}
		return NULL;
	}
	else
	{
		return NULL;
	}
};

std::vector<CoreNative::ImageCube*>* CoreNative::CubeManager::GetAllImageCubes()
{
	return cubes;
};

void CoreNative::CubeManager::RemoveImageCube(const char* imageID)
{
	if(!cubes->empty())
	{
		for( std::vector<CoreNative::ImageCube*>::iterator iterator = cubes->begin();
			iterator!= cubes->end(); iterator++)
		{
			if( strncmp(imageID,(*iterator)->GetImageID(), strlen(imageID)) == 0 )
			{
				cubes->erase(iterator);
				break;
			}
		}
	}
};

CoreNative::ImageCube* CoreNative::CubeManager::CreateImageCube(const char* hedFile, const char* ctxFile, const char* id)
{
	CoreNative::ImageCube* cube = reader->ReadFiles(hedFile,ctxFile,id);
	cubes->push_back(cube);
	return cube;
};

const char* CoreNative::CubeManager::GetImageCubeByFilename(const char* filename)
{
	if(!cubes->empty())
	{
		for(unsigned int i=0; i<cubes->size(); i++)
		{
			CoreNative::ImageCube* tempCube = cubes->at(i);
			if( strncmp(filename,tempCube->GetHedFile(), strlen(filename)) == 0 )
			{
				return tempCube->GetImageID();
			}
		}
		return NULL;
	}
	else
	{
		return NULL;
	}
};

File: trunk/CT-Viewer/Viewer/Core/CubeManager.cs
using System;
using System.Collections.Generic;

namespace Viewer.Core
{
    public class CubeManager
    {
        private readonly Dictionary<Guid, ImageCube> imageCubes;

        public CubeManager()
        {
            imageCubes = new Dictionary<Guid, ImageCube>();
        }

        public ImageCube GetImageCube(Guid imageID)
        {
            return imageCubes[imageID];
        }

        public void RemoveImageCube(Guid imageID)
        {
            imageCubes.Remove(imageID);
        }

        public ImageCube CreateImageCube(string hedFile, string ctxFile)
        {
            ImageCube cube = Reader.ReadFiles(hedFile, ctxFile);
            imageCubes.Add(cube.ImageID,cube);
            return cube;
        }

        public Guid GetImageCubeByFilename(string filename)
        {
            Guid returnValue = Guid.Empty;
            foreach (KeyValuePair<Guid,ImageCube> entry in imageCubes)
            {
                if(entry.Value.HedFile.CompareTo(filename) == 0 ||
                   entry.Value.CtxFile.CompareTo(filename) == 0 )
                {
                    returnValue = entry.Key;
                }
            }
            return returnValue;
        }
    }
}